src/services/ftp/backend.rs

```diff
@@ -0,0 +1,440 @@
+            .finish()
+    }
+}
+
+impl Backend {
+    pub(crate) fn from_iter(it: impl Iterator<Item = (String, String)>) -> Result<Self> {
+        let mut builder = Builder::default();
+
+        for (k, v) in it {
+            let v = v.as_str();
+            match k.as_ref() {
+                "root" => builder.root(v),
+                _ => continue,
+            };
+        }
+        builder.build()
+    }
+}
+
+#[async_trait]
+impl Accessor for Backend {
+    async fn create(&self, args: &OpCreate) -> Result<()> {
+        let path = args.path();
+        if args.mode() == ObjectMode::FILE {
+            //let mut ftp_stream = self.ftp_connect()?;
+            self.client
+                .try_lock()
+                .unwrap()
+                .put_file(&path, &mut "".as_bytes())
+                .map_err(|e| {
+                    other(ObjectError::new(
+                        Operation::Create,
+                        path,
+                        anyhow!("put request: {e:?}"),
+                    ))
+                })?;
+            return Ok(());
+        }
+
+        if args.mode() == ObjectMode::DIR {
+            self.client.try_lock().unwrap().mkdir(&path).map_err(|e| {
+                other(ObjectError::new(
+                    Operation::Create,
+                    path,
+                    anyhow!("mkdir request: {e:?}"),
+                ))
+            })?;
+
+            return Ok(());
+        }
+
+        unreachable!()
+    }
+
+    async fn read(&self, args: &OpRead) -> Result<BytesReader> {
+        let path = args.path();
+        let mut guard = self.client.lock().await;
+
+        let mut stream = guard.retr_as_stream(path).map_err(|e| {
+            other(ObjectError::new(
+                Operation::Read,
+                path,
+                anyhow!("retrieve request: {e:?}"),
+            ))
+        })?;
+
+        if let Some(offset) = args.offset() {
+            copy(&mut stream.by_ref().take(offset), &mut sink())?;
+        }
+
+        let mut buf = Vec::new();
+        let r = match args.size() {
+            None => {
+                stream.by_ref().read_to_end(&mut buf)?;
+                Box::new(Cursor::new(buf))
+            }
+            Some(size) => {
+                stream.by_ref().take(size).read_to_end(&mut buf)?;
+                Box::new(Cursor::new(buf))
+            }
+        };
+
+        guard.finalize_retr_stream(stream).map_err(|e| {
+            other(ObjectError::new(
+                Operation::Read,
+                path,
+                anyhow!("finalizing stream request: {e:?}"),
+            ))
+        })?;
+
+        drop(guard);
+
+        Ok(r)
+    }
+
+    async fn write(&self, args: &OpWrite, r: BytesReader) -> Result<u64> {
+        let path = args.path();
+
+        let mut guard = self.client.lock().await;
+
+        let mut reader = Box::pin(r);
+
+        let mut buf = Vec::new();
+
+        let _byte = reader.read_to_end(&mut buf).await?;
```

Comment:
Is it better to use https://docs.rs/suppaftp/latest/suppaftp/struct.FtpStream.html#method.put_with_stream?

Reply:
Oh, I got it. We are using sync API, which doesn't accept `BytesReader` here.